File: Package.swift
// swift-tools-version: 6.2
//===----------------------------------------------------------------------===//
// Copyright © 2026 container-compose project authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//===----------------------------------------------------------------------===//

import Foundation
import PackageDescription

let runtimeProfile = ProcessInfo.processInfo.environment["CONTAINER_COMPOSE_BUILD_PROFILE"] ?? "enhanced"
let enhancedRuntime: Bool = {
    switch runtimeProfile {
    case "enhanced": true
    case "stock": false
    default: fatalError("CONTAINER_COMPOSE_BUILD_PROFILE must be stock or enhanced")
    }
}()

let runtimeSwiftSettings: [SwiftSetting] = enhancedRuntime
    ? [.define("CONTAINER_COMPOSE_ENHANCED_RUNTIME")]
    : []

let containerDependency: Package.Dependency = {
    if let path = ProcessInfo.processInfo.environment["CONTAINER_PACKAGE_PATH"],
       !path.isEmpty
    {
        return .package(name: "container", path: path)
    }
    return enhancedRuntime
        ? .package(
            url: "https://github.com/stephenlclarke/container.git",
            revision: "ccf99d73b75626ed49a0d638c640bd3b9851e2de",
        )
        : .package(url: "https://github.com/apple/container.git", exact: "1.4.1")
}()

let containerizationDependency: Package.Dependency = {
    if let path = ProcessInfo.processInfo.environment["CONTAINERIZATION_PACKAGE_PATH"],
       !path.isEmpty
    {
        return .package(name: "containerization", path: path)
    }
    return enhancedRuntime
        ? .package(
            url: "https://github.com/stephenlclarke/containerization.git",
            revision: "bd8130fea851f6ee264f00fc684e2543a7d2faa3",
        )
        : .package(url: "https://github.com/apple/containerization.git", exact: "0.45.0")
}()

let runtimeOnlyDependencies: [Package.Dependency] = enhancedRuntime
    ? [.package(
        url: "https://github.com/stephenlclarke/swift-nio-ssl.git",
        revision: "3e13ce5f6dd5b7e89fff9ab55ab7caed39fe7285",
    )]
    : []

let pluginRuntimeDependencies: [Target.Dependency] = enhancedRuntime
    ? ["ComposeContainerRuntime"]
    : ["ComposeEngineRuntime"]

let runtimeTargets: [Target] = enhancedRuntime
    ? [
        .target(
            name: "ComposeContainerRuntime",
            dependencies: [
                .product(name: "ArgumentParser", package: "swift-argument-parser"),
                "ComposeCore",
                "ComposeRuntimeSPI",
                .product(name: "ContainerAPIClient", package: "container"),
                .product(name: "ContainerCommands", package: "container"),
                .product(name: "ContainerPersistence", package: "container"),
                .product(name: "ContainerResource", package: "container"),
                .product(name: "Containerization", package: "containerization"),
                .product(name: "ContainerizationArchive", package: "containerization"),
                .product(name: "ContainerizationEXT4", package: "containerization"),
                .product(name: "ContainerizationExtras", package: "containerization"),
                .product(name: "ContainerizationOCI", package: "containerization"),
                .product(name: "Logging", package: "swift-log"),
            ],
            path: "Sources/ComposeContainerRuntime",
            swiftSettings: runtimeSwiftSettings,
        ),
        .testTarget(
            name: "ComposeCoreTests",
            dependencies: [
                "ComposeCore",
                "ComposeContainerRuntime",
                .product(name: "ContainerResource", package: "container"),
                .product(name: "ContainerizationArchive", package: "containerization"),
                .product(name: "ContainerizationExtras", package: "containerization"),
            ],
            path: "Tests/ComposeCoreTests",
            resources: [
                .process("Fixtures"),
            ],
        ),
        .testTarget(
            name: "ComposeContainerRuntimeTests",
            dependencies: [
                "ComposeContainerRuntime",
                "ComposeRuntimeSPI",
                .product(name: "ContainerResource", package: "container"),
                .product(name: "ContainerizationEXT4", package: "containerization"),
                .product(name: "ContainerizationOCI", package: "containerization"),
            ],
            path: "Tests/ComposeContainerRuntimeTests",
        ),
    ]
    : [
        .target(
            name: "ComposeEngineRuntime",
            dependencies: [
                "ComposeCore",
                "ComposeRuntimeSPI",
                .product(name: "ContainerEngineWire", package: "container-engine-api"),
                .product(name: "ContainerUnixHTTPClient", package: "container-engine-api"),
            ],
            path: "Sources/ComposeEngineRuntime",
        ),
        .testTarget(
            name: "ComposeEngineRuntimeTests",
            dependencies: [
                "ComposeEngineRuntime",
                "ComposeRuntimeSPI",
                .product(name: "ContainerEngineWire", package: "container-engine-api"),
                .product(name: "ContainerUnixHTTPServer", package: "container-engine-api"),
                .product(name: "Logging", package: "swift-log"),
            ],
            path: "Tests/ComposeEngineRuntimeTests",
        ),
    ]

let package = Package(
    name: "container-compose",
    platforms: [.macOS(.v15)],
    products: [
        .executable(name: "compose", targets: ["ComposePlugin"]),
        .library(name: "ComposeCore", targets: ["ComposeCore"]),
        enhancedRuntime
            ? .library(name: "ComposeContainerRuntime", targets: ["ComposeContainerRuntime"])
            : .library(name: "ComposeEngineRuntime", targets: ["ComposeEngineRuntime"]),
        .library(name: "ComposeRuntimeSPI", targets: ["ComposeRuntimeSPI"]),
    ],
    dependencies: [
        containerDependency,
        containerizationDependency,
        .package(url: "https://github.com/apple/swift-argument-parser.git", from: "1.3.0"),
        .package(url: "https://github.com/swiftlang/swift-docc-plugin.git", from: "1.4.0"),
        .package(url: "https://github.com/apple/swift-log.git", from: "1.0.0"),
        .package(
            url: "https://github.com/stephenlclarke/container-engine-api.git",
            revision: "276a7cfdba91fef60c232177a44c054e5de9ae8f",
        ),
    ] + runtimeOnlyDependencies,
    targets: [
        .executableTarget(
            name: "ComposePlugin",
            dependencies: [
                .product(name: "ArgumentParser", package: "swift-argument-parser"),
                .product(name: "ContainerAPIClient", package: "container"),
                .product(name: "ContainerBuild", package: "container"),
                .product(name: "ContainerCommands", package: "container"),
                .product(name: "ContainerLog", package: "container"),
                .product(name: "ContainerResource", package: "container"),
                "ComposeCore",
            ] + pluginRuntimeDependencies,
            path: "Sources/ComposePlugin",
            swiftSettings: runtimeSwiftSettings,
        ),
        .target(
            name: "ComposeRuntimeSPI",
            path: "Sources/ComposeRuntimeSPI",
        ),
        .target(
            name: "ComposeCore",
            dependencies: [
                "ComposeRuntimeSPI",
            ],
            path: "Sources/ComposeCore",
        ),
        .testTarget(
            name: "ComposeRuntimeSPITests",
            dependencies: [
                "ComposeRuntimeSPI",
            ],
            path: "Tests/ComposeRuntimeSPITests",
        ),
        .testTarget(
            name: "ComposePluginTests",
            dependencies: [
                "ComposeCore",
                "ComposePlugin",
            ],
            path: "Tests/ComposePluginTests",
            swiftSettings: runtimeSwiftSettings,
        ),
        .testTarget(
            name: "ComposeRuntimeTests",
            dependencies: [
                "ComposeCore",
            ],
            path: "Tests/ComposeRuntimeTests",
            resources: [
                .copy("Fixtures"),
            ],
        ),
    ] + runtimeTargets,
)

File: Tools/bazel/package.py
#!/usr/bin/env python3
##===----------------------------------------------------------------------===##
## Copyright © 2026 container-compose project authors.
##
## Licensed under the Apache License, Version 2.0 (the "License");
## you may not use this file except in compliance with the License.
## You may obtain a copy of the License at
##
##   https://www.apache.org/licenses/LICENSE-2.0
##
## Unless required by applicable law or agreed to in writing, software
## distributed under the License is distributed on an "AS IS" BASIS,
## WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
## See the License for the specific language governing permissions and
## limitations under the License.
##===----------------------------------------------------------------------===##

"""Validate declared native products and write unsigned candidate provenance."""

from __future__ import annotations

import argparse
import hashlib
import importlib.util
import json
from pathlib import Path
import re
import struct


PRODUCTS = {
    "compose": "darwin-arm64",
    "compose-normalizer": "darwin-arm64",
    "compose-volume-initializer-linux-arm64": "linux-arm64",
    "compose-volume-initializer-linux-amd64": "linux-amd64",
}


def sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as stream:
        for chunk in iter(lambda: stream.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def validate_binary(path: Path, platform: str) -> None:
    """Reject wrong architectures and dynamically linked guest initializers."""
    with path.open("rb") as stream:
        header = stream.read(64)
        if platform == "darwin-arm64":
            if header[:8] != struct.pack("<II", 0xFEEDFACF, 0x0100000C):
                raise ValueError("Expected a thin arm64 Mach-O product")
            return
        machine = {"linux-arm64": 183, "linux-amd64": 62}[platform]
        if (len(header) != 64 or header[:7] != b"\x7fELF\x02\x01\x01"
                or struct.unpack_from("<HH", header, 16) != (2, machine)):
            raise ValueError("Expected a static ELF executable for the declared architecture")
        offset = struct.unpack_from("<Q", header, 32)[0]
        size, count = struct.unpack_from("<HH", header, 54)
        if size != 56 or count < 1 or count > 1024 or offset < 64:
            raise ValueError("Invalid ELF program header inventory")
        stream.seek(offset)
        for _ in range(count):
            record = stream.read(size)
            if len(record) != size or struct.unpack_from("<I", record)[0] in {2, 3}:
                raise ValueError("Guest initializer is truncated or dynamically linked")


def dependency(pins: list[dict], name: str, profile: str) -> tuple[str, str]:
    """Use only the selected immutable lock; never discover installed runtimes."""
    matches = [pin for pin in pins if pin["identity"] == name]
    if len(matches) != 1:
        raise ValueError("Missing or duplicate runtime dependency")
    pin = matches[0]
    source = re.fullmatch(r"https://github.com/([A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+)\.git", pin["location"])
    revision = pin["state"]["revision"]
    if not source or not re.fullmatch(r"[0-9a-f]{40}", revision):
        raise ValueError("Runtime dependency must have a GitHub source and immutable revision")
    if profile == "stock" and source[1] != "apple/" + name:
        raise ValueError("Stock candidate must use unmodified Apple dependency pins")
    return source[1], revision


def metadata(manifest: dict, writer: Path) -> tuple[dict, dict]:
    """Reuse existing build-info semantics without invoking builds or signing."""
    if set(manifest["binaries"]) != set(PRODUCTS):
        raise ValueError("Candidate needs exactly four declared products")
    if manifest["profile"] not in {"stock", "enhanced"} or not re.fullmatch(r"[0-9a-f]{40}", manifest["commit"]):
        raise ValueError("Candidate requires a selected profile and full source commit")
    version = re.findall(r"^COMPOSE_VERSION\s*\?=\s*([0-9]+\.[0-9]+\.[0-9]+)\s*$", Path(manifest["makefile"]).read_text(), re.M)
    go_version = re.findall(r"^\s*github.com/compose-spec/compose-go/v2\s+(v[0-9]+\.[0-9]+\.[0-9]+)\s*$", Path(manifest["go_mod"]).read_text(), re.M)
    if len(version) != 1 or len(go_version) != 1:
        raise ValueError("Missing unique project or compose-go version")
    pins = json.loads(Path(manifest["resolved"]).read_bytes())["pins"]
    container_source, container_ref = dependency(pins, "container", manifest["profile"])
    virtualization_source, virtualization_ref = dependency(pins, "containerization", manifest["profile"])
    capabilities, schema = [], 1
    if manifest["profile"] == "enhanced":
        spec = importlib.util.spec_from_file_location("compose_build_info", writer)
        module = importlib.util.module_from_spec(spec)
        spec.loader.exec_module(module)
        schema, capabilities = module.load_runtime_capability_manifest(Path(manifest["capabilities"]))
    info = {
        "version": version[0], "source": "stephenlclarke/container-compose",
        "branch": "detached", "lane": "candidate", "commit": manifest["commit"],
        "buildType": "release", "containerSource": container_source, "containerRef": container_ref,
        "containerizationSource": virtualization_source, "containerizationRef": virtualization_ref,
        "composeGoVersion": go_version[0], "runtimeCapabilitySchemaVersion": schema,
        "runtimeCapabilities": capabilities,
    }
    products = {}
    for name, platform in PRODUCTS.items():
        path = Path(manifest["binaries"][name])
        validate_binary(path, platform)
        products[name] = sha256(path)
    identity = {
        "schemaVersion": 1, "kind": "unsigned-native-candidate", "productFamily": "container-compose",
        "version": version[0], "commit": manifest["commit"], "runtimeProfile": manifest["profile"],
        "architecture": "arm64", "compilationMode": "opt", "distributionReady": False,
        "licenseClosureComplete": False, "products": products,
        "dependencyLockSHA256": sha256(Path(manifest["resolved"])),
        "goModSHA256": sha256(Path(manifest["go_mod"])), "goSumSHA256": sha256(Path(manifest["go_sum"])),
    }
    return info, identity


def receipt(archive: Path, identity: dict) -> dict:
    """Seal exact archive bytes; a receipt is not Developer ID or release proof."""
    return {**identity, "archiveSHA256": sha256(archive), "archiveSize": archive.stat().st_size}


def write_json(path: Path, value: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(value, indent=2, sort_keys=True) + "\n")


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    commands = parser.add_subparsers(dest="command", required=True)
    build = commands.add_parser("metadata")
    for name in ("manifest", "info", "identity", "writer"):
        build.add_argument(name, type=Path)
    seal = commands.add_parser("receipt")
    for name in ("archive", "identity", "output"):
        seal.add_argument(name, type=Path)
    args = parser.parse_args()
    if args.command == "metadata":
        info, identity = metadata(json.loads(args.manifest.read_bytes()), args.writer)
        write_json(args.info, info)
        write_json(args.identity, identity)
    else:
        write_json(args.output, receipt(args.archive, json.loads(args.identity.read_bytes())))


if __name__ == "__main__":
    main()

